### DISCUSSION TITLE
Gauge with no value?

### DISCUSSION THREAD

**Discussion by endotronic** [prometheus/client_python#731] [General]
I think of a gauge as a momentary reading of something. A good example would be a value from a sensor. For this, imagine a thermometer. A gauge would be a good representation of the temperature from that thermometer. Well, what if the thermometer becomes unreadable? In this situation I would like to stop reporting the temperature at all, since it is now unknown. Keeping the last value set would misrepresent the situation and hide a condition that could be used for alerting.

I really like to use a callback with gauges  (`set_function`) since this reflects how Prometheus works (e.g. read the thermometer when Prometheus scrapes the endpoint). It seems natural to raise an exception in this situation, if a value cannot be determined, but I find that this causes all metrics (even working ones) to go missing, as the HTTP endpoint begins to return "A server error occurred.  Please contact the administrator." It should be possible to let one metric disappear while keeping others that still work intact.

Even using the gauge the "normal" way, I'm not sure how one would go about removing it. I imagine you cannot set the value to None, since returning None in the callback I just mentioned results in `TypeError: float() argument must be a string or a number, not 'NoneType'`.

Is there a way to temporarily remove a gauge?

**Comment by csmarchbanks** [comment]
Hello, the best solution is probably to use a [Custom Collector](https://github.com/prometheus/client_python#custom-collectors) to gather the temperature information and only create a gauge if there is a value. This keeps the behavior of reading the thermometer when Prometheus scrapes the endpoint, and allows you lots of control over the results.

It might also be possible to remove the gauge as needed, e.g. `registry.unregister(temperature_gauge)`. Then re-register it when the value comes back. Keeping track of this state is probably more of a pain than using the custom collector though.

**Reply by endotronic** [reply]
Ah, Custom Collector looks great, thanks!

**Reply by jvoosten** [reply]
Related to this is the Gauge initialization problem, I guess. Consider the following:

* I want to report the values of a digital electricity & gas meter using the P1 port.
* In the constructor of my code I create the Gauges for the various measurements.
* Then the main loop starts. It may take up to 20 seconds before the first data comes in.
* In the mean time the Prometheus server scrapes the data, all zeroes.
* First real data is parsed, Gauges are set() with their values. 
* Prometheus scrapes again and stores some huge numbers like 60000 kWh of electricty.

The result: giant spikes in the output graphs that make the data unusable. No process runs 24/7 forever so these gaps are expected.

So what can we do?
- The "obvious" solution would be to create the gauges after the first real data has come in but that's just plain ugly coding; constructors are there for a reason.
- The workaround in a query "if it is 0 it doesn't count" is exactly that: a workaround. 
- Registering and unregistering gauges on the fly is a possibility but unnecessary if your set of gauges is fixed anyway and only have a very short 'zero time'. Besides, there is a subtle difference between a gauge not being listed, or a gauge being there but not having a value.
- Custom Collectors makes the code more complicated for for just a simple value.

Conclusion: there really should be a way to set a metric (gauge, summary, etc) to "No value (yet)".

**Reply by csmarchbanks** [reply]
I think this is a reasonable request, if custom collectors are too hard to use (and that seems to be the case, or at least they are not obvious to use) we should fix that. I can see a couple options:
1. Make custom collectors easier to use, maybe a small abstraction to make it very easy for simple values?
2. Support None for `set_function`. I am coming around to this as it is basically a custom collector anyway. Not quite as versatile as 1 since we could potentially support more types.

**Reply by jvoosten** [reply]
I think allowing None is the simplest and most straightforward. When I see the word "Custom Collector" I think of code that has to do complicated things or values that cannot/should not be collected in a simple loop. At the very least you need to create an object per variable, which seems overkill for what's essentially a float.

Regarding types: I can't think of anything else than None that would indicate 'no value'. In Python that's basically its whole purpose...
